### PR TITLE
TargetTypes by Upgrade and Replacement of Targetable* by Targetable and ITargetablePositions

### LINKED ISSUE #8724: Proposal to refactor TargetType handling
Currently we have like 4-5 traits which handle TargetTypes on actors.
TargetableBuilding returns the cells
TargetableUnit works on units
and multiple subclasses of TargetableUnit, all working around the fact that an actor can have only one Targetable* trait.

I honestly didn't really looked into the code much, but from what I saw, merging TargetableBuilding and TargetableUnit is possible (via moving the Building() check in the latter) and then instead of having TargetableAircraft etc, we could set up their source traits granting an upgrade to enable/disable targeting/affection. (Helicopter granting an upgrade at height 200 for example)

This will be important for the RA2 mod because at there unit-targeting could change on the fly (Temporal weapons made the target only affected by such, an already mindcontrolled unit cannot be remindcontrolled, IronCurtained units weren't attacked by the AI, units in-air (paradropped or flying) couldn't get targeted by strictly ground weapons, etc).

This would also render map scripting easier.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/24400080-proposal-to-refactor-targettype-handling?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by atlimit8**
Closes #8724, unless there is something else to it.

### CODE REVIEW

**Comment by atlimit8** [comment]
Closing for replacement.

**Comment by atlimit8** [comment]
Note: the first commit is being replaced - different system.

**Comment by GraionDilach** [comment]
Cannot this be extended to TargetableBuilding as well?

**Comment by atlimit8** [comment]
I simply missed that one, but plan on including it now.

**Comment by pchote** [comment]
See my point (1) in #8724.  Merging TargetableBuilding into this is asking for trouble.

**Comment by atlimit8** [comment]
My plan is create ITargetablePositions and have Mobile, Building, and Aircraft implement it.

**Comment by atlimit8** [comment]
I know the TargetTypes is updated by upgrade changes. I used Console.WriteLine to double check. However, I'm not sure the targeting system always recognizes the changes.

Please check whether landing aircraft has an effect on targetability.

File: OpenRA.Mods.Common/Activities/Air/HeliFly.cs [hunk 1]
```diff
@@ -53,6 +53,9 @@ public override Activity Tick(Actor self)
 			if (IsCanceled || !target.IsValidFor(self))
 				return NextActivity;
 
+			if (helicopter.CenterPosition.Z == 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/HeliFly.cs]: #8866 introduced an `IsAtGroundLevel()` that works correctly for mods that use heightmaps.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 2]
```diff
@@ -188,6 +192,8 @@ public void AddedToWorld(Actor self)
 			self.World.ActorMap.AddInfluence(self, this);
 			self.World.ActorMap.AddPosition(self, this);
 			self.World.ScreenMap.Add(self);
+			if (CenterPosition.Z > 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: These z checks should be replaced with IsAtGroundLevel too.

**Comment by atlimit8** [comment]
Manages airborne upgrades via Aircraft.IsAirborne property now.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 3]
```diff
@@ -172,6 +204,7 @@ public void SetPosition(Actor self, WPos pos)
 			{
 				self.World.ScreenMap.Update(self);
 				self.World.ActorMap.UpdatePosition(self, this);
+				IsAirborne = self.World.Map.DistanceAboveTerrain(CenterPosition).Length != 0;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: `> 0`, just in the off chance that someone makes a subterranean aircraft that doesn't break before this.

**Comment by atlimit8** [comment]
Added MinAirborneAltitude to Aircraft trait to set airborne minimum altitude.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by atlimit8** [comment]
Rebased.

**Comment by abcdefg30** [comment]
Looks good to me otherwise. :+1: / :white_check_mark: 

As a (nice) side-effect you can't force fire at the prisoners in allies03 anymore.

**Comment by atlimit8** [comment]
Changed.

**Comment by Mailaender** [comment]
Needs a rebase.

**Comment by abcdefg30** [comment]
Needs a rebase again.

**Comment by atlimit8** [comment]
Rebased.

**Comment by abcdefg30** [comment]
Needs a rebase again.

**Comment by atlimit8** [comment]
Rebased.
It should be noted that this PR replaces the prisoner cloak hack in the Allies 03 missions but without causing #8722 like #8662 did.

**Comment by atlimit8** [comment]
Rebased again.
Thanks #9236.

**Comment by Mailaender** [comment]
Aircraft and submarine targeting still works. I can also confirm the fix for Allies 03. :white_check_mark:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2cfc5c677533d94d8b2a47e8b419da945dd8740c...2f356e99955907302f98381ac31fd4856fa215f1)

**Pull request merged**